app/liquid_tags/glitch_tag.rb

```diff
@@ -31,49 +37,70 @@ def render(_context)
 
   private
 
-  def valid_id?(input)
-    (input =~ ID_REGEXP)&.zero?
+  def parsed_input(input)
+    id, *options = input.split
+    stripped_id = id.gsub(/^~/, "") # remove possible preceeding tilde
+    match = pattern_match_for(stripped_id, REGEXP_OPTIONS)
+    raise StandardError, I18n.t("liquid_tags.glitch_tag.invalid_glitch_id") unless match
+
+    [get_slug(match), parse_options(options, match)]
   end
 
-  def parse_id(input)
-    id = input.split.first
-    id.sub!(TILDE_PREFIX_REGEXP, "")
-    raise StandardError, I18n.t("liquid_tags.glitch_tag.invalid_glitch_id") unless valid_id?(id)
+  def get_slug(match)
+    if match_has_named_capture_group?(match, "subdomain")
+      match[:subdomain].presence
+    else
+      match[:slug]
+    end
+  end
 
-    id
+  def match_has_named_capture_group?(match, group_name)
+    match.names.include?(group_name)
   end
 
-  def valid_option(option)
-    option.match(OPTION_REGEXP)
+  def parse_options(options, match)
+    # 'app' and 'code' should cancel each other out
+    options -= %w[app code] if (options & %w[app code]) == %w[app code]
+
+    # check for file= in options or path= in params; fallback is file=index.html
+    file_option = options.detect { |option| option.start_with?("file=") }
+    options += ["file=#{path_within_params(match)}"] unless file_option
+
+    validated_options = options.select { |option| valid_option?(option) }
+    raise StandardError, I18n.t("liquid_tags.glitch_tag.invalid_options") if validated_options.empty?
+
+    build_options(validated_options)
+  end
+
+  def path_within_params(match)
+    return "index.html" unless match_has_named_capture_group?(match, "params")
+
+    return "index.html" if match[:params].blank?
+
+    path_match = pattern_match_for(match[:params], [PATH_REGEX])
+    return "index.html" if path_match.blank?
+
+    path_match[:path]
+  end
+
+  def valid_option?(option)
+    option.match?(OPTION_REGEXP)
   end
 
   def build_options(options)
     # Convert options to query param pairs
     params = options.filter_map { |option| OPTIONS_TO_QUERY_PAIR[option] }
 
-    # Deal with the file option if present or use default
+    # by this point, there is always a file_option
     file_option = options.detect { |option| option.start_with?("file=") }
-    path = file_option ? (file_option.sub! "file=", "") : "index.html"
+    path = (file_option.sub! "file=", "")
```

Comment:
I know this isn't the case for this specific usage here, but `String#sub!` can return `nil`, so we shouldn't use it like this. We have a couple different options here (unfortunately, GitHub won't let me do a 1-click diff suggestion because of the deleted line in the diff).

The first option is to use `sub` instead of `sub!`:

```ruby
    file_option = options.detect { |option| option.start_with?("file=") }
    path = file_option.sub("file=", "")
```

… or this could even be another use case for `String#delete_prefix`:

```ruby
    path = options
      .detect { |option| option.start_with?("file=") }
      .delete_prefix("file=")
```

The other is to avoid the intermediate `file_option` variable entirely:

```ruby
    path = options.detect { |option| option.start_with?("file=") }
    path.sub! "file=", ""
```

What follows is completely optional information, not required for this review but I'm adding it because it could be useful for understanding why someone likely chose this and why it may be a good idea to move away from it:

The first involves an additional Ruby heap allocation, so _technically_ the second is faster, which is often why people use `sub!` or `gsub!` over the non-`!` version. That's a premature optimization, though. The difference in performance is on the order of 1µs and is only noticeable (or even measurable, really) in a hot loop or some other aggregate. As a basis of comparison, a single DB query takes thousands of times as long even in the best-case scenario and allocates dozens to hundreds of times more memory.

Also, I notice that this `sub!` call could be mutating strings created outside this object. Mutating values you received but did not create yourself is a common source of bugs, so it's probably a good idea to avoid the in-place string mutation anyway.